pkg/controller/job/job_controller.go

```diff
@@ -545,6 +555,16 @@ func (jm *Controller) enqueueSyncJobInternal(logger klog.Logger, obj interface{}
 		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", obj, err))
 		return
 	}
+
+	// This is a performance optimization. When the Job object is managed by
+	// an external controller we don't need to queue the synchronization tasks.
+	if jobObj, ok := obj.(*batch.Job); ok {
+		if controllerName := managedByExternalController(jobObj); controllerName != nil {
+			logger.V(2).Info("Skipped syncing of the job as it is managed by an external controller", "key", key, "uid", jobObj.UID, "controllerName", controllerName)
```

Comment:
This is logged on every request to queue. 

However, so is https://github.com/kubernetes/kubernetes/blob/e457683e6cb740656f389932fb564a6ab45be759/pkg/controller/job/job_controller.go#L554. So, we will get the same amount of logging as if it was reconciled by the main controller. Having this information logged may be helpful for debugging as probably `v=2` is the default logging level for most prod environments. 

Let me know what you think.  

For now I remove completely v(2) from code to make it clear the intention is to be at the same logging level as if it was queued.

Reply:
As I understand it, this targets mainly the multi-cluster use case and we could end up with a larger amount of jobs than usual. I think we should not log jobs that are a responsibility of other controllers which now should take care of the logging.

For people debugging this, they will first see that their jobs are not progressing and have the following options:
1. look at the managed-by label
2. look at the metrics (set up alerts)
3. increase the log level of the controller, which is not an uncommon debugging step

Reply:
Definitely not `logger.Info` but `V(2)` and higher is fine by me. I have a different question. Why do we need this skip in both the enqueuing logic and in sync? Shouldn't the former be sufficient?

Reply:
Moved logging to v(2). The check inside `syncJob` is needed because even if we do the skip in the enqueuing logic it is still possible we perform the sync. The scenario:
1. job x created without managed-by label, and reconciliation is queued
2. job x is deleted and a new job with the same key is created
3. the previously queued reconciliation of the job x starts, and status is updated and pods

This scenario is possible e2e, and thus I added the `TestManagedByLabel_RecreatedJob` integration test, replicating this issue, if we only rely on the skip in the enqueuing logic.

There are two more scenarios identified, which maybe are preventable, but since we have the check inside anyway it does not seem necessary:
1. if the job is deleted and we get the `DeletedFinalStateUnknown` object instead of Job obj (I'm not sure how to reliably replicate it for testing), we could get the Job obj from the  `DeletedFinalStateUnknown` instance, but it does not seem to be guaranteed there (as there are cases when we have `deletedObj=nil`: https://github.com/kubernetes/kubernetes/blob/a67973a45c4b48585e3331889eca09425caca7c2/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L681-L692
2. job changing the `spec.ActiveDeadlineSeconds` - this is preventable for sure: https://github.com/kubernetes/kubernetes/blob/210378e918a225611719772ab728c127a4e51cd6/pkg/controller/job/job_controller.go#L482-L489